### PR TITLE
Add sample for GPT2 inference

### PR DESCRIPTION
**Pull request by Abdurrahheem**
### Pull Request Readiness Checklist

This PR adds sample for inferencing GPT-2 model. More specificly implementation of GPT-2 from [this repository](https://github.com/karpathy/build-nanogpt). Currently inference in OpenCV is only possible to do with fixed window size due to not supported dynamic shapes. 

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/gpt2_inference.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+'''
+This is a sample script to run GPT-2 inference in OpenCV using ONNX model.
+The script loads the GPT-2 model and runs inference on a given prompt.
+Curently script only works with fixed size window, that means
+you will have to specify promt of the same lenth as when model was exported to ONNX.
+
+
+Exporting GPT-2 model to ONNX.
+To expoert GPT-2 model to ONNX, you can use the following preocedure:
```
Comment (asmorkalov) [inline samples/dnn/gpt2_inference.py:9]: typo

File: samples/dnn/gpt2_inference.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        # Get the probabilities using softmax
+        probs = stable_softmax(logits)
+
+        # Do top-k sampling of 50
+        topk_indices = np.argpartition(probs, -50, axis=-1)[:, -50:]
+        topk_probs = np.take_along_axis(probs, topk_indices, axis=-1)
+
+        # Normalize top-k probabilities
+        topk_probs /= np.sum(topk_probs, axis=-1, keepdims=True)
```
Comment (dkurt) [inline samples/dnn/gpt2_inference.py:75]: Do we really need softmax and normalization to just get top-k predictions? I think their order is not changed from raw logits.

File: samples/dnn/gpt2_inference.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    python gpt2_inference.py --model=<path-to-onnx-model> --max_seq_len=<max-output-lenght> --batch_size=<use-one-used-while-exportinh> --prompt=<use-promt-of-the-same-length-used-while-exporting>
+'''
+
+
+
+from copy import deepcopy
+import numpy as np
+import tiktoken
+import argparse
+import cv2
```
Comment (dkurt) [inline samples/dnn/gpt2_inference.py:35]: `import cv2 as cv`

File: samples/dnn/gpt2_inference.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    args = parse_args()
+    max_length = args.max_seq_len
+    num_return_sequences = args.batch_size
+    prompt = args.prompt
+
+    enc = tiktoken.get_encoding('gpt2')
+    tokens = enc.encode(prompt)
+
+    net = cv2.dnn.readNetFromONNX(args.model)
```
Comment (dkurt) [inline samples/dnn/gpt2_inference.py:102]: maybe just `readNet`?

File: samples/dnn/gpt2_inference.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+Run the script:
+
+    python gpt2_inference.py --model=<path-to-onnx-model> --max_seq_len=<max-output-lenght> --batch_size=<use-one-used-while-exportinh> --prompt=<use-promt-of-the-same-length-used-while-exporting>
+'''
+
+
+
+from copy import deepcopy
+import numpy as np
+import tiktoken
```
Comment (dkurt) [inline samples/dnn/gpt2_inference.py:33]: External dependency. Add instruction of how to install and which version.

File: samples/dnn/gpt2_inference.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    pip install -r requirements.txt
+
+3  Export the model to ONNX:
+
+    python export2onnx.py --promt=<Any-promt-you-want> --batch_size=<batch-size>
+
+
+Run the script:
+
+    python gpt2_inference.py --model=<path-to-onnx-model> --max_seq_len=<max-output-lenght> --batch_size=<use-one-used-while-exportinh> --prompt=<use-promt-of-the-same-length-used-while-exporting>
```
Comment (dkurt) [inline samples/dnn/gpt2_inference.py:26]: > --prompt=<use-promt-of-the-same-length-used-while-exporting>

Can sample support simple padding in case of shorter prompt? Otherwise user cannot understand how to fill the same length.

Comment (Abdurrahheem) [reply samples/dnn/gpt2_inference.py:26]: Currenly no

Comment (Abdurrahheem) [reply samples/dnn/gpt2_inference.py:26]: how to know head of time which prompt legth user used?

Comment (dkurt) [reply samples/dnn/gpt2_inference.py:26]: If model exported with some input shape we can accept prompt with less or equal length and pad in case of shorter prompt. There is an option to obtain input shape in OpenCV like this: https://github.com/opencv/opencv/blob/6a11847d579e95bc5cca64d951684d4b578870ae/modules/dnn/src/model.cpp#L51

File: samples/dnn/gpt2_inference.py (hunk 2 above)
Comment (Abdurrahheem) [reply samples/dnn/gpt2_inference.py:75]: `np.random.choice` expect probabilities. otherwise you are correct

File: samples/dnn/gpt2_inference.py (hunk 6 above)
Comment (dkurt) [reply samples/dnn/gpt2_inference.py:26]: Let's use `max_input_length` as an additional parameter: https://github.com/search?q=repo%3Ahuggingface%2Ftransformers%20max_input_length&type=code

File: samples/dnn/gpt2_inference.py [hunk 7]
```diff
@@ -96,10 +101,22 @@ def gpt2_inference(net, tokens, max_length, num_return_sequences=5):
...
 
-    net = cv2.dnn.readNetFromONNX(args.model)
+    # Check if the prompt is of the same length as the input tokens
+    # if not, pad the tokens else truncate the tokens
+    if len(tokens) > input_token_size:
+        tokens = tokens[:input_token_size]
+    elif len(tokens) < input_token_size:
+        tokens2pad = input_token_size - len(tokens)
+        # append <space> token to the prompt
+        tokens += [220] * tokens2pad
```
Comment (dkurt) [inline samples/dnn/gpt2_inference.py:117]: Not sure that whitespace is a correct pad token. There are also versions with EOS or PAD. Bit ok to merge with space now.

source: https://github.com/search?q=repo%3Ahuggingface%2Ftransformers+pad_token&type=code&p=2

**Comment by asmorkalov** [comment]
Looks like the new CV_WARP breaks Java bindings:
```
In file included from modules/java_bindings_generator/gen/cpp/opencv_jni.hpp:4,
                 from /home/ci/opencv/modules/java/generator/src/cpp/opencv_java.cpp:8:
/home/ci/build/modules/java_bindings_generator/gen/cpp/dnn.inl.hpp: In function 'void Java_org_opencv_dnn_Net_getLayerShapes_10(JNIEnv*, jclass, jlong, jlong, jint, jint, jobject, jobject)':
/home/ci/build/modules/java_bindings_generator/gen/cpp/dnn.inl.hpp:4924:9: error: 'Copy_vector_MatShape_to_List' was not declared in this scope; did you mean 'Copy_vector_string_to_List'?
 4924 |         Copy_vector_MatShape_to_List(env,inLayerShapes,inLayerShapes_list);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         Copy_vector_string_to_List
```

**Comment by Abdurrahheem** [comment]
> Looks like the new CV_WARP breaks Java bindings:
> 
> ```
> In file included from modules/java_bindings_generator/gen/cpp/opencv_jni.hpp:4,
>                  from /home/ci/opencv/modules/java/generator/src/cpp/opencv_java.cpp:8:
> /home/ci/build/modules/java_bindings_generator/gen/cpp/dnn.inl.hpp: In function 'void Java_org_opencv_dnn_Net_getLayerShapes_10(JNIEnv*, jclass, jlong, jlong, jint, jint, jobject, jobject)':
> /home/ci/build/modules/java_bindings_generator/gen/cpp/dnn.inl.hpp:4924:9: error: 'Copy_vector_MatShape_to_List' was not declared in this scope; did you mean 'Copy_vector_string_to_List'?
>  4924 |         Copy_vector_MatShape_to_List(env,inLayerShapes,inLayerShapes_list);
>       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
>       |         Copy_vector_string_to_List
> ```

Is it possible to some how resolve this? 
The functionality of `net.getLayerShapes` quite usefull in case (when we do not have dynamisc yet)

**Comment by asmorkalov** [comment]
Give me several minutes, I'll mush solution to the PR.

File: samples/dnn/gpt2_inference.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+3  Export the model to ONNX:
+
+    python export2onnx.py --promt=<Any-promt-you-want> --batch_size=<batch-size>
+
+
+Run the script:
+1. Install the required dependencies:
+
+    pip install tiktoken==0.7.0 numpy==1.23.0
```
Comment (asmorkalov) [inline samples/dnn/gpt2_inference.py:27]: You can remove numpy from the list. It's already there as OpenCV dependency.

File: samples/dnn/gpt2_inference.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        x = x[:, 1:] ## issue due to fixes size window in opencv
+
+        output_buffer = np.concatenate((output_buffer, sampled_indices), axis=1)
+        counter += 1
+    print("Inference done!")
+    return output_buffer
+
+if __name__ == '__main__':
+
+    np.random.seed(0)
```
Comment (asmorkalov) [inline samples/dnn/gpt2_inference.py:97]: It makes sense to use command line parameter for seed. Zero is good default value.

**Review by asmorkalov — APPROVED** [review]

File: samples/dnn/gpt2_inference.py (hunk 7 above)
Comment (Abdurrahheem) [reply samples/dnn/gpt2_inference.py:117]: EOS I tried this. Your prompt followed by EOS make model output sentences that are not related to your prompt (as you indicatited end of context by EOS)

File: modules/java/generator/src/cpp/listconverters.cpp [hunk 10]
```diff
@@ -108,3 +108,21 @@ void Copy_vector_string_to_List(JNIEnv* env, std::vector<std::string>& vs, jobje
...
+    env->CallVoidMethod(list, m_clear);
+    for (std::vector<cv::dnn::MatShape>::iterator it = vs.begin(); it != vs.end(); ++it)
+    {
+        jintArray element = env->NewIntArray((int)it->size());
+        env->SetIntArrayRegion(element, 0, (int)it->size(), &(*it)[0]);
+        env->CallBooleanMethod(list, m_add, element);
+        env->DeleteLocalRef(element);
+    }
+}
+#endif // HAVE_OPENCV_DNN
```
Comment (dkurt) [inline modules/java/generator/src/cpp/listconverters.cpp:128]: @asmorkalov, 
```
  listconverters.cpp
     Creating library C:/build/precommit_windows64/build/lib/Release/opencv_xstereo500.lib and object C:/build/precommit_windows64/build/lib/Release/opencv_xstereo500.exp
C:\build\precommit_windows64\5.x\opencv\modules\java\generator\src\cpp\listconverters.cpp(123): error C2664: 'void JNIEnv_::SetIntArrayRegion(jintArray,jsize,jsize,const jint *)': cannot convert argument 4 from 'int *' to 'const jint *' [C:\build\precommit_windows64\build\modules\java\jni\opencv_java.vcxproj]
  C:\build\precommit_windows64\5.x\opencv\modules\java\generator\src\cpp\listconverters.cpp(123): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
  opencv_java.cpp
  ```

Comment (asmorkalov) [reply modules/java/generator/src/cpp/listconverters.cpp:128]: Fixed.

**Review by dkurt — APPROVED** [review]
